File: api/internal/metrics/metrics.go
package metrics

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	PredictionsTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "deepstream_predictions_total",
			Help: "Total number of predictions triggered.",
		},
		[]string{"status"},
	)

	InferenceLatency = promauto.NewHistogram(
		prometheus.HistogramOpts{
			Name:    "deepstream_inference_latency_ms",
			Help:    "Inference latency in milliseconds.",
			Buckets: []float64{1, 2, 5, 10, 20, 50, 100},
		},
	)

	ActiveWebsockets = promauto.NewGauge(
		prometheus.GaugeOpts{
			Name: "deepstream_active_websockets",
			Help: "Current active WebSocket connections.",
		},
	)

	HTTPRequestsTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "deepstream_http_requests_total",
			Help: "Total number of HTTP requests.",
		},
		[]string{"method", "path", "status"},
	)
)

File: inference/src/metrics.rs
use lazy_static::lazy_static;
use prometheus::{opts, register_histogram, register_int_counter, Histogram, IntCounter};

lazy_static! {
    pub static ref INFERENCE_REQUESTS: IntCounter = register_int_counter!(
        opts!(
            "deepstream_inference_requests_total",
            "Total number of inference requests"
        )
    )
    .unwrap();
    pub static ref INFERENCE_LATENCY_HISTOGRAM: Histogram = register_histogram!(
        "deepstream_inference_latency_ms",
        "Inference latency in milliseconds",
        vec![1.0, 2.0, 5.0, 10.0, 20.0, 50.0, 100.0]
    )
    .unwrap();
}
